cirq/circuits/frozen_circuit.py

```diff
@@ -0,0 +1,166 @@
+                 device: 'cirq.Device' = devices.UNCONSTRAINED_DEVICE) -> None:
+        base = Circuit(contents, strategy=strategy, device=device)
+        self._moments = tuple(base.moments)
+        self._device = base.device
+
+        # These variables are memoized when first requested.
+        self._num_qubits = None
+        self._qid_shape = None
+        self._all_qubits = None
+        self._all_operations = None
+        self._unitary = None
+        self._has_measurements = None
+        self._all_measurement_keys = None
+        self._are_all_measurements_terminal = None
+
+    @property
+    def moments(self) -> Sequence['cirq.Moment']:
+        return self._moments
+
+    @property
+    def device(self) -> devices.Device:
+        return self._device
+
+    def __hash__(self):
+        return hash((self.moments, self.device))
+
+    # Memoized methods for commonly-retrieved properties.
+
+    def _num_qubits_(self):
+        if self._num_qubits is None:
+            self._num_qubits = len(self.all_qubits())
+        return self._num_qubits
+
+    def _qid_shape_(self):
+        if self._qid_shape is None:
+            self._qid_shape = super()._qid_shape_()
+        return self._qid_shape
+
+    def _unitary_(self):
+        if self._unitary is None:
+            self._unitary = super()._unitary_()
+        return self._unitary
+
+    def all_qubits(self):
+        if self._all_qubits is None:
+            self._all_qubits = super().all_qubits()
+        return self._all_qubits
+
+    def all_operations(self):
+        if self._all_operations is None:
+            self._all_operations = tuple(super().all_operations())
+        return self._all_operations
+
+    def has_measurements(self):
+        if self._has_measurements is None:
+            self._has_measurements = super().has_measurements()
+        return self._has_measurements
+
+    def all_measurement_keys(self):
+        if self._all_measurement_keys is None:
+            self._all_measurement_keys = super().all_measurement_keys()
+        return self._all_measurement_keys
+
+    def are_all_measurements_terminal(self):
+        if self._are_all_measurements_terminal is None:
+            self._are_all_measurements_terminal = super(
+            ).are_all_measurements_terminal()
+        return self._are_all_measurements_terminal
+
+    # End of memoized methods.
+
+    def __add__(self, other):
+        return (self.unfreeze() + other).freeze()
+
+    def __radd__(self, other):
+        return (other + self.unfreeze()).freeze()
+
+    # TODO: handle multiplication / powers differently?
+    def __mul__(self, other):
+        return (self.unfreeze() * other).freeze()
+
+    def __rmul__(self, other):
+        return (other * self.unfreeze()).freeze()
+
+    def __pow__(self, other):
+        try:
+            return (self.unfreeze()**other).freeze()
+        except:
+            return NotImplemented
+
+    # pylint: disable=function-redefined
```

Comment:
Don't need this lint unless you are going to actually redefine the function to specify type overloads. I think it'd be a good idea to add the overloads.

Reply:
IIRC, mypy complains about these if I include the overloads: `Overloaded method has both abstract and real implementations`, or something along those lines. I'll remove the pylint-disable comment, though.